telegraf.go

```diff
@@ -60,313 +61,115 @@ type TelegrafConfigFilter struct {
-			Type:    p.Config.Type(),
-			Comment: p.Comment,
-			Config:  p.Config,
-		}
-	}
-	return json.Marshal(tce)
-}
-
-// UnmarshalTOML implements toml.Unmarshaler interface.
-func (tc *TelegrafConfig) UnmarshalTOML(data interface{}) error {
-	dataOk, ok := data.(map[string]interface{})
-	if !ok {
-		return errors.New("blank string")
-	}
-	agent, ok := dataOk["agent"].(map[string]interface{})
-	if !ok {
-		return errors.New("agent is missing")
-	}
-
-	intervalStr, ok := agent["interval"].(string)
-	if !ok {
-		return errors.New("agent interval is not string")
-	}
-
-	interval, err := time.ParseDuration(intervalStr)
-	if err != nil {
-		return err
-	}
-	tc.Agent = TelegrafAgentConfig{
-		Interval: interval.Nanoseconds() / 1000000,
-	}
-
-	for tp, ps := range dataOk {
-		if tp == "agent" {
-			continue
-		}
-		plugins, ok := ps.(map[string]interface{})
-		if !ok {
-			return &Error{
-				Msg: "bad plugin type",
-			}
-		}
-		for name, configDataArray := range plugins {
-			if configDataArray == nil {
-				if err := tc.parseTOMLPluginConfig(tp, name, configDataArray); err != nil {
-					return err
-				}
-				continue
-			}
-			for _, configData := range configDataArray.([]map[string]interface{}) {
-				if err := tc.parseTOMLPluginConfig(tp, name, configData); err != nil {
-					return err
-				}
-			}
-		}
-	}
-
-	return nil
-}
-
-func (tc *TelegrafConfig) parseTOMLPluginConfig(typ, name string, configData interface{}) error {
-	var ok bool
-	var tpFn func() plugins.Config
-	switch typ {
-	case "inputs":
-		tpFn, ok = availableInputPlugins[name]
-	case "outputs":
-		tpFn, ok = availableOutputPlugins[name]
-	default:
-		return &Error{
-			Msg: fmt.Sprintf(ErrUnsupportTelegrafPluginType, typ),
-		}
-	}
-
-	if !ok {
-		return &Error{
-			Msg: fmt.Sprintf(ErrUnsupportTelegrafPluginName, name, typ),
-		}
-	}
-	p := tpFn()
-
-	if err := p.UnmarshalTOML(configData); err != nil {
-		return err
-	}
-	tc.Plugins = append(tc.Plugins, TelegrafPlugin{
-		Config: p,
-	})
-	return nil
+	ID          ID     `json:"id,omitempty"`          // ID of this config object.
+	OrgID       ID     `json:"orgID,omitempty"`       // OrgID is the id of the owning organization.
+	Name        string `json:"name,omitempty"`        // Name of this config object.
+	Description string `json:"description,omitempty"` // Decription of this config object.
+	Config      string `json:"config,omitempty"`      // ConfigTOML contains the raw toml config.
 }
 
 // UnmarshalJSON implement the json.Unmarshaler interface.
+// Gets called when reading from the kv db. mostly legacy so loading old/stored configs still work.
+// May not remove for a while. Primarily will get hit when user views/downloads config.
 func (tc *TelegrafConfig) UnmarshalJSON(b []byte) error {
 	tcd := new(telegrafConfigDecode)
 	if err := json.Unmarshal(b, tcd); err != nil {
 		return err
 	}
+
 	orgID := tcd.OrgID
 	if !orgID.Valid() {
 		orgID = tcd.OrganizationID
 	}
+
 	*tc = TelegrafConfig{
```

Comment:
can you assign these values straight to the TelegrafConfig instead of deref/assing combo? somethign like

```go
tc.ID = tcd.ID
tc.OrgID = orgID
... yada yada
```

Reply:
ah ya.. what about `tc = &TelegrafConfig{` instead?

Reply:
I'm fairly certain that will create an issue where the call site won't see the changes, b/c hte pointer value itself changed internally to the method 🤔, at least that's what intelliJ always yells at me 🤣